### PR TITLE
feat: Added IsMeasurement option to Column attribute

### PR DESCRIPTION
**Pull request by vesuvian**
## Proposed Changes

Added IsMeasurement option to Column attribute, allowing for dynamic measurement names in POCO classes

## Checklist

- [x] CHANGELOG.md updated
- [x] Rebased/mergeable
- [x] A test has been added if appropriate
- [x] `dotnet test` completes successfully
- [x] Commit messages are in [semantic format](https://seesparkbox.com/foundry/semantic_commit_messages)
- [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

### CODE REVIEW

**Comment by vesuvian** [comment]
I've been able to run the most immediately relevant unit tests without issue, but I'm struggling to run tests that depend on a live instance of InfluxDB. How can I configure the tests to run against my InfluxDB server?

**Comment by bednar** [comment]
Thanks for your PR 👍 

> I've been able to run the most immediately relevant unit tests without issue, but I'm struggling to run tests that depend on a live instance of InfluxDB. How can I configure the tests to run against my InfluxDB server?


You can configure following env variables:

1. InfluxDB 1.8+ - `INFLUXDB_IP`, `INFLUXDB_PORT_API`
2. InfluxDB 2 - `INFLUXDB_2_IP`, `INFLUXDB_2_PORT`

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,7 @@ Adds a `Type` overload for POCOs to `QueryAsync`. This will add `object ConvertT
 ### Features
 1. [#232](https://github.com/influxdata/influxdb-client-csharp/pull/232): Add a `Type` overload for POCOs to `QueryAsync`.
 1. [#233](https://github.com/influxdata/influxdb-client-csharp/pull/233): Add possibility to follow HTTP redirects
+1. [#240](https://github.com/influxdata/influxdb-client-csharp/pull/240): Add IsMeasurement option to Column attribute for dynamic measurement names in POCO classes
```
Comment (bednar) [inline CHANGELOG.md:11]: Please, move it to release `3.1.0`.

File: Client.Linq/Internal/Expressions/MeasurementColumnName.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        }
+
+        public void AppendFlux(StringBuilder builder)
+        {
+            builder.Append("r[\"");
+            _delegate.AppendFlux(builder);
+            builder.Append("\"]");
+        }
+    }
+}
```
Comment (bednar) [inline Client.Linq/Internal/Expressions/MeasurementColumnName.cs:22]: Please, add test for filtering by measurement in LINQ driver

https://github.com/influxdata/influxdb-client-csharp/blob/a671601c727072a4d07466d04748ff9b16b52888/Client.Linq.Test/InfluxDBQueryVisitorTest.cs

**Review by bednar — CHANGES_REQUESTED** [review]
Thanks again for your PR 👍 

There are a few requirements that must be be satisfy before we accept the PR:

**Comment by vesuvian** [comment]
@bednar I am having a miserable time trying to get these tests to run. I've been able to set up my environment variables, I've enabled HTTP in my influxdb.conf, and I am running an instance of InfluxDB 2.0.8

- Legacy tests fail because of authorization, so I have temporarily hardcoded adding my token to the auth header
- Now legacy tests fail with a response from the server: "not implemented: CREATE DATABASE" - ok, they're legacy, so I guess I can skip them for v2?
- LINQ tests that depend on port 9999 immediately fail, so I change it to 8086
- LINQ tests now fail because ItInfluxDBQueryableTest.SetUp() hardcodes the token as "my-token" - why? should this be an environment variable? Am I meant to do a find and replace?

Is there some documentation I'm missing for how to correctly set up a test environment?

**Comment by bednar** [comment]
> @bednar I am having a miserable time trying to get these tests to run. I've been able to set up my environment variables, I've enabled HTTP in my influxdb.conf, and I am running an instance of InfluxDB 2.0.8
> 
> * Legacy tests fail because of authorization, so I have temporarily hardcoded adding my token to the auth header
> * Now legacy tests fail with a response from the server: "not implemented: CREATE DATABASE" - ok, they're legacy, so I guess I can skip them for v2?
> * LINQ tests that depend on port 9999 immediately fail, so I change it to 8086
> * LINQ tests now fail because ItInfluxDBQueryableTest.SetUp() hardcodes the token as "my-token" - why? should this be an environment variable? Am I meant to do a find and replace?
> 
> Is there some documentation I'm missing for how to correctly set up a test environment?

The best way to prepare testing environment is use a `Scripts/influxdb-restart.sh` script. It will start InfluxDBs and also correctly setup authorization.

File: Client/Internal/MeasurementMapper.cs [hunk 3]
```diff
@@ -35,18 +35,30 @@ internal PointData ToPoint<TM>(TM measurement, WritePrecision precision)
 
             var measurementType = measurement.GetType();
             CacheMeasurementClass(measurementType);
-            
+
             var measurementAttribute = (Measurement) measurementType.GetCustomAttribute(typeof(Measurement));
-            if (measurementAttribute == null)
+            var measurementColumn = CACHE[measurementType.Name].SingleOrDefault(p => p.Column.IsMeasurement);
+
+            if ((measurementAttribute == null) ^ (measurementColumn == null) == false)
```
Comment (bednar) [inline Client/Internal/MeasurementMapper.cs:42]: Please change expression to, for better maintenance and understanding by other users:

```suggestion
            if (((measurementAttribute == null) ^ (measurementColumn == null)) == false)
```

File: Client.Linq/Internal/Expressions/MeasurementColumnName.cs (hunk 2 above)
Comment (bednar) [reply Client.Linq/Internal/Expressions/MeasurementColumnName.cs:22]: Something like will be pretty fine:

```csharp
public void FilterByMeasurement()
{
    var query = from s in InfluxDBQueryable<SensorWithCustomMeasurement>.Queryable("my-bucket", "my-org", _queryApi)
        where s.Value > 10
        where s.Measurement == "my-measurement"
        select s;
    var visitor = BuildQueryVisitor(query);

    const string expected = "start_shifted = int(v: time(v: p2))\n\nfrom(bucket: p1) " +
                            "|> range(start: time(v: start_shifted)) " +
                            "|> filter(fn: (r) => (r[\"_measurement\"] == p4)) " +
                            "|> pivot(rowKey:[\"_time\"], columnKey: [\"_field\"], valueColumn: \"_value\") " +
                            "|> drop(columns: [\"_start\", \"_stop\", \"_measurement\"]) " +
                            "|> filter(fn: (r) => (r[\"data\"] > p3))";

    Assert.AreEqual(expected, visitor.BuildFluxQuery());
    
    var ast = visitor.BuildFluxAST();
    
    var measurementAssignment = ((OptionStatement) ast.Body[3]).Assignment as VariableAssignment;
    Assert.AreEqual("p4", measurementAssignment?.Id.Name);
    Assert.AreEqual("my-measurement", (measurementAssignment?.Init as StringLiteral)?.Value);
}
```

```csharp
public class SensorWithCustomMeasurement
{
    [Column(IsMeasurement = true)]
    public string Measurement { get; set; }

    [Column("sensor_id", IsTag = true)] 
    public string SensorId { get; set; }

    [Column("data")]
    public int Value { get; set; }
}
```

You have to update a [`QueryVisitor`](https://github.com/influxdata/influxdb-client-csharp/blob/25e9f7cf5404274c961442af95f6b876ce4c1b3b/Client.Linq/Internal/QueryVisitor.cs#L88) by:

image

**Review by bednar — CHANGES_REQUESTED** [review]

**Comment by vesuvian** [comment]
@bednar thank you so much for the hand holding. I was ultimately able to set up a debian container with docker, the .net SDK, this repository, and run the influxdb-restart and ci-test.sh scripts with no errors.

I hope you don't mind but I took the liberty of:
- Removed duplicate entry for PR 239 from the changelog
- AbstractTest clarifies HTTP request failures, consistent with AbstractTest.WaitToCallback
- ItTasksApiTest.Runs compares against system time as UTC, previously this test was failing for me
- ci-test.sh now uses `dotnet tool update` instead of `dotnet tool install` - install would previously fail with error code 1 on subsequent runs https://github.com/dotnet/sdk/issues/9500

Something still isn't quite right when using linq. My results appear to be correct but the measurement property is not being populated. I'll keep digging and see what I can find.

**Comment by vesuvian** [comment]
The measurement column was being explicitly dropped by the QueryAggregator in BuildFluxQuery. I have removed _measurement from the list of dropped columns.

Is this ok by you? I don't want to introduce any performance issues.

**Comment by bednar** [comment]
> The measurement column was being explicitly dropped by the QueryAggregator in BuildFluxQuery. I have removed _measurement from the list of dropped columns.
> 
> Is this ok by you? I don't want to introduce any performance issues.

It is definitely increase amount of data transferred from the server. What do you think about configure this dropping by [QueryableOptimizerSettings](https://github.com/influxdata/influxdb-client-csharp/blob/533c46fd701bcb552fb86494462a8f6116157182/Client.Linq/InfluxDBQueryable.cs#L16)?

**Comment by vesuvian** [comment]
I've added optimization settings for dropping start, stop and measurement columns.

At this point I believe I'm done. My application is happily reading/writing to influx using both raw flux and linq queries.

The failing test `ItWriteQueryApiTest.Jitter()` looks to be entirely unrelated, I'm wondering if there may be a race condition with the timings? I've been able to successfully run all of the tests on my machine.

**Comment by bednar** [comment]
> The failing test `ItWriteQueryApiTest.Jitter()` looks to be entirely unrelated, I'm wondering if there may be a race condition with the timings?

I don’t think so. 

It looks like that `ItWriteQueryApiTest.Jitter()` is little bit flaky.

**Review by bednar — CHANGES_REQUESTED** [review]
Thanks again for your PR 👍. We are almost done.

Can you please update a README.md for LINQ driver with information how to use filtering by measurement?

https://github.com/influxdata/influxdb-client-csharp/blob/master/Client.Linq/README.md

**Comment by vesuvian** [comment]
@bednar I've gone ahead and updated the README documents, please let me know if you need anything else

**Review by bednar — APPROVED** [review]
LGTM 👍

**Pull request merged**